url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+		    
+		    // the maximum size component
+		    Set<Integer> greatestComponent = null; // TODO testar tempos com isto
+		    
+		    // which nodes have already been used (either in visited or in toVisit)
+		    Set<Integer> used = new HashSet<Integer>();
+		    
+		    // iterates the nodes in the graph
+		    Iterator<Map.Entry<Integer, Set<Integer>>> it = graph.entrySet().iterator();
+		    
+		    // let's iterate through the nodes in the graph, looking for the maximum
+		    // size component. we will be doing a breadth first seach // TODO optimize this with a better algorithm?
+		    while (it.hasNext() && !found) {
+			// get our initial node
+			Map.Entry<Integer, Set<Integer>> node = it.next();
+			int initialNode = node.getKey();
+			
+			// we can't have used it before!
+			if ( used.contains(initialNode) )
+			    continue;
+			
+			// what are the edges of our initial node?
+			Integer[] edges = (Integer[]) node.getValue().toArray(new Integer[1]);
+			
+			// allocate memory for the set of visited nodes
+			visited = new HashSet<Integer>();
+			
+			// the set of nodes we must still visit
+			List<Integer> toVisit = new LinkedList<Integer>();
+			
+			// add the initial node to the set of used and visited nodes
+			visited.add(initialNode);
+			used.add(initialNode);
+			
+			// add my edges to the set of nodes we must visit
+			// and also put them in the used set
+			for (Integer edge : edges) {
+			    toVisit.add(edge);
+			    used.add(edge);
+			}
+			
+			// start the search!
+			while (toVisit.size() != 0) {
+			    // the node we are visiting
+			    int no = toVisit.remove(0);
+			    
+			    // add node to visited set
+			    visited.add(no);
+			    
+			    // queue edges to be visited (if they haven't been already
+			    for (Integer edge : graph.get(no))
+				if ( !visited.contains(edge) )
+				    toVisit.add(edge);
+			}
+			
+			// is the number of visited nodes, greater than the 'currently' largest component?
+			if ( visited.size() > maximumSize ) { // it is! we've found a greater component then...
+			
+			    // update the maximum size
+			    maximumSize = visited.size();
+			    
+			    // update our greatest component
+			    greatestComponent = visited;
+			} else
+			    continue;
+		    }
+		    
+		    /*
+		     * we've found the maximum size connected component -- 'greatestComponent'
+		     */
+		    
+		    // let's choose the row
+		    for (Row row : rows.values()) {
+			// is it a node in the graph?
+			if ( row.edges != null ) { // it is
+			
+			    // get the nodes connected through this edge
+			    Integer[] edge = row.edges.toArray(new Integer[2]);
+			    int node1 = edge[0];
+			    int node2 = edge[1];
+			    
+			    // is this row an edge in the maximum size component?
+			    if ( greatestComponent.contains(node1) && greatestComponent.contains(node2) ) {
+				rLinha = row.id;
+				break;
+			    } else
+				continue;
+			} else
+			    continue;
+		    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1476:1493, 2361:2376.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.